File: client/src/lib/normalization.ts
/**
 * Normalizes article codes for fuzzy matching
 * 1. Convert to uppercase
 * 2. Remove delimiters (space, -, _, ., /)
 * 3. Replace Cyrillic characters with Latin equivalents
 */
export function normalizeArticle(article: string): string {
  if (!article) return '';
  
  // Step 1: Convert to uppercase
  let normalized = article.toUpperCase();
  
  // Step 2: Remove delimiters
  normalized = normalized.replace(/[\s\-_./]/g, '');
  
  // Step 3: Replace Cyrillic with Latin
  const cyrillicToLatin: { [key: string]: string } = {
    'А': 'A',
    'В': 'B',
    'Е': 'E',
    'К': 'K',
    'М': 'M',
    'Н': 'H',
    'О': 'O',
    'Р': 'P',
    'С': 'C',
    'Т': 'T',
    'У': 'Y',
    'Х': 'X',
    'Ё': 'E',
  };
  
  normalized = normalized.replace(/[АВЕКМНОРСТУХЁ]/g, (char) => {
    return cyrillicToLatin[char] || char;
  });
  
  return normalized;
}

/**
 * Checks if two articles match after normalization
 */
export function articlesMatch(article1: string, article2: string): boolean {
  return normalizeArticle(article1) === normalizeArticle(article2);
}

File: py_backend/normalization.py
"""Article normalization logic ported from shared/normalization.ts."""

from __future__ import annotations

import re

_CYRILLIC_TO_LATIN: dict[str, str] = {
    "А": "A",
    "В": "B",
    "Е": "E",
    "К": "K",
    "М": "M",
    "Н": "H",
    "О": "O",
    "Р": "P",
    "С": "C",
    "Т": "T",
    "У": "Y",
    "Х": "X",
    "Ё": "E",
}


def normalize_article(article: str) -> str:
    """Normalize article codes for fuzzy matching."""
    if not article:
        return ""

    normalized = article.upper()
    normalized = re.sub(r"[\s\-_./]", "", normalized)
    normalized = re.sub(
        r"[АВЕКМНОРСТУХЁ]",
        lambda m: _CYRILLIC_TO_LATIN.get(m.group(0), m.group(0)),
        normalized,
    )
    return normalized


def articles_match(article1: str, article2: str) -> bool:
    """Check whether two article codes match after normalization."""
    return normalize_article(article1) == normalize_article(article2)


def normalize_box_name(name: str) -> str:
    """Normalize box/storage location names to avoid duplicates/typos.

    Rules (aligned with article normalization where it makes sense):
    - uppercase
    - remove common delimiters (space, -, _, ., /)
    - replace Cyrillic lookalikes with Latin (АВЕКМНОРСТУХЁ)
    """
    if not name:
        return ""

    normalized = name.strip().upper()
    normalized = re.sub(r"[\s\-_./]", "", normalized)
    normalized = re.sub(
        r"[АВЕКМНОРСТУХЁ]",
        lambda m: _CYRILLIC_TO_LATIN.get(m.group(0), m.group(0)),
        normalized,
    )
    return normalized


def box_names_match(name1: str, name2: str) -> bool:
    return normalize_box_name(name1) == normalize_box_name(name2)


# TypeScript-compatible aliases.
normalizeArticle = normalize_article
articlesMatch = articles_match
normalizeBoxName = normalize_box_name
boxNamesMatch = box_names_match
